include/ibamr/AdvDiffConservativeCUIConvectiveOperator.h

```diff
@@ -0,0 +1,221 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2018 - 2020 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDE GUARD ////////////////////////////////
+
+#ifndef included_IBAMR_AdvDiffConservativeCUIConvectiveOperator
+#define included_IBAMR_AdvDiffConservativeCUIConvectiveOperator
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibamr/config.h>
+
+#include "ibamr/ConvectiveOperator.h"
+#include "ibamr/ibamr_enums.h"
+
+#include "ibtk/ibtk_utilities.h"
+
+#include "CellVariable.h"
+#include "CoarsenAlgorithm.h"
+#include "FaceVariable.h"
+#include "IntVector.h"
+#include "PatchHierarchy.h"
+#include "RefineAlgorithm.h"
+#include "RefinePatchStrategy.h"
+#include "tbox/Database.h"
+#include "tbox/Pointer.h"
+
+#include <string>
+#include <vector>
+
+namespace SAMRAI
+{
+namespace solv
+{
+template <int DIM, class TYPE>
+class SAMRAIVectorReal;
+template <int DIM>
+class RobinBcCoefStrategy;
+} // namespace solv
+namespace xfer
+{
+template <int DIM>
+class CoarsenSchedule;
+template <int DIM>
+class RefineSchedule;
+} // namespace xfer
+} // namespace SAMRAI
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBAMR
+{
+/*!
+ * \brief Class AdvDiffConservativeCUIConvectiveOperator is a concrete ConvectiveOperator
```

Comment:
How does this class differ from `AdvDiffCUIConvectiveOperator`, which also is supposed to have a "convective" mode (determined by the `ConvectiveDifferencingType` specified in the class constructor)?

The name `AdvDiffConservativeCUIConvectiveOperator` makes it sound like this is a specialization of `AdvDiffCUIConvectiveOperator`. Is that right?

Reply:
Yes. `AdvDiffConservativeCUIConvectiveOperator` class is a specialization of `AdvDiffCUIConvectiveOperator` class. Because this class also uses `CUI` as a convective operator type.  But the difference is `AdvDiffConservativeCUIConvectiveOperator` class computes \div{u Q1 Q2} whereas `AdvDiffCUIConvectiveOperator` class computes \div{u Q} or u.\grad{Q} based on the ConvectiveDifferencingType.

Reply:
I think this class may need a new name.

We do not have a 100% consistent naming system, but `AdvDiffConservativeCUIConvectiveOperator` suggests that it is a specialization (in the C++/object oriented programming sense) of `AdvDiffCUIConvectiveOperator`, but the header indicates that it is actually a specialization of class `ConvectiveOperator`. 

In fact, it seems like its main use case (to compute terms of the form `N = div(Q1 Q2 u)` doesn't really fit with the `ConvectiveOperator` interface. (`applyConvectiveOperator(int Q1_idx, int Q2_idx, int N_idx)` is only provided by this implementation.)

I haven't looked into the implementation, but does this basically just apply the standard CUI scheme with `Q = Q1*Q2`? If so, would suggest:

1. Provide an implementation of `applyConvectiveOperator(int Q1_idx, int Q2_idx, int N_idx)` in the base class in terms of the virtual method `applyConvectiveOperator(int Q_idx, int N_idx)`.
2. Provide an adaptor class that can use any `ConvectiveOperator` implementation to evaluate `applyConvectiveOperator(int Q1_idx, int Q2_idx, int N_idx)`.

The current design will require duplicating every `ConvectiveOperator` to also provide this other form, which I would prefer to avoid if possible.

Reply:
@boyceg The class name is `AdvDiffCUIConservativeConvectiveOperator`. Since this class is a specialization of `AdvDiffCUIConvectiveOperator` it is derived from the `AdvDiffCUIConvectiveOperator` class. This class's main use is to compute `N = div(Q P u)`.

I have added a virtual function `applyConvectiveOperator(int Q_idx, int P_idx, int N_idx)` in the base class `ConvectiveOperator`. But the implementation is provided only in `AdvDiffCUIConservativeConvectiveOperator`.

Please take a look and let me know your suggestions.

Reply:
I think we need to come up with a more precise name for the concept that is being provided here. Calling it a "conservative" specialization of the parent operator seems misleading, since the original CUI operator implementation itself also can be set up to perform conservative differencing.

From an implementation standpoint, it might be better to split out the cell-to-face flux computation and synchronization from the conservative differencing operation (which itself could be done in the base class) --- that way any convective operator implementation could be used to perform these operations without requiring a custom specialization with lots of duplicated code.

Reply:
> From an implementation standpoint, it might be better to split out the cell-to-face flux computation and synchronization from the conservative differencing operation (which itself could be done in the base class) --- that way any convective operator implementation could be used to perform these operations without requiring a custom specialization with lots of duplicated code.

I think this is really the right thing to do --- i.e. splitting out the functionality into separate functions. I think it will reduce the amount of duplicated code between the different operator implementations. I will try to put up an implementation in the next day or two...